controllers/reclaimspacejob_controller.go

```diff
@@ -18,45 +18,420 @@ package controllers
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the ReclaimSpaceJob object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.10.0/pkg/reconcile
 func (r *ReclaimSpaceJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	// Fetch ReclaimSpaceJob instance.
+	rsJob := &csiaddonsv1alpha1.ReclaimSpaceJob{}
+	err := r.Client.Get(ctx, req.NamespacedName, rsJob)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			logger.Info("ReclaimSpaceJob resource not found")
+
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	if !rsJob.DeletionTimestamp.IsZero() {
+		logger.Info("ReclaimSpaceJob resource is being deleted, exiting reconcile")
+		return ctrl.Result{}, nil
+	}
+
+	if rsJob.Status.Result != "" {
+		logger.Info(fmt.Sprintf("ReclaimSpaceJob is already in %q state, exiting reconcile",
+			rsJob.Status.Result))
+		// since result is already set, just dequeue.
+		return ctrl.Result{}, nil
+	}
+
+	err = validateReclaimSpaceJobSpec(rsJob)
+	if err != nil {
+		logger.Error(err, "Failed to validate ReclaimSpaceJob.Spec")
+
+		rsJob.Status.Result = csiaddonsv1alpha1.OperationResultFailed
+		rsJob.Status.Message = fmt.Sprintf("Failed to validate ReclaimSpaceJob.Spec: %v", err)
+		rsJob.Status.CompletionTime = &v1.Time{Time: time.Now()}
+		if statusErr := r.Client.Status().Update(ctx, rsJob); statusErr != nil {
+			logger.Error(err, "Failed to update status")
+			return ctrl.Result{}, statusErr
+		}
+
+		// invalid parameters, do not requeue.
+		return ctrl.Result{}, nil
+	}
+
+	// set default values if equal to 0.
+	if rsJob.Spec.BackoffLimit == 0 {
+		rsJob.Spec.BackoffLimit = defaultBackoffLimit
+	}
+	if rsJob.Spec.RetryDeadlineSeconds == 0 {
+		rsJob.Spec.RetryDeadlineSeconds = defaultRetryDeadlineSeconds
+	}
+
+	err = r.reconcile(
+		ctx,
+		&logger,
+		rsJob,
+		req.Namespace,
+	)
+
+	if rsJob.Status.Result == "" && rsJob.Status.Retries == rsJob.Spec.BackoffLimit {
+		logger.Info("Maximum retry limit reached")
+		rsJob.Status.Result = csiaddonsv1alpha1.OperationResultFailed
+		rsJob.Status.Message = "Maximum retry limit reached"
+		rsJob.Status.CompletionTime = &v1.Time{Time: time.Now()}
+	}
+
+	if statusErr := r.Client.Status().Update(ctx, rsJob); statusErr != nil {
+		logger.Error(statusErr, "Failed to update status", "reclaimSpaceJob", rsJob)
 
-	// TODO(user): your logic here
+		return ctrl.Result{}, statusErr
+	}
 
-	return ctrl.Result{}, nil
+	if rsJob.Status.Result != "" {
```

Comment:
> when this can happen? because we are already doing return if the status.Result is set.

That check is before calling `reconcile()`, the result maybe set inside `reconcile` function and status updated, therefore we do not need to re-queue it. 

Any err returned by reconcile without `status.Result` being set is treated as intermediate/temporary error and requeue happens.